### PR TITLE
Greedy Follower

### PR DESCRIPTION
**Pull request by erikwijmans**
## Motivation and Context

The current action space shortest implementation is un-reliable and slow :(

Having the ability to generate a path in action space serves two purposes:
1.  Videos of traveling between two points
1. An oracle to learn from

The action space shortest path code serves neither of these purposes as it's slow and unreliable.  This PR provides a _much_ faster and reliable greedy follower.  While it doesn't generate the shortest path in action space, the paths still tend to be very good.

This needs to add some logging, @msavva we should figure out how to have unified logging between c++ and python.  Should we just bind glog?

## How Has This Been Tested

This has been tested on every navmesh I could find.  584 of them in in total.  Of the ~600k source/target pairs in the testing, 94 of them failed, which I consider to be very, very good.

The failures cases that I looked at tended to be in poor reconstructions where oddities in geometry led to very narrow choke points that the agent is unable to travel down due to not having continuous rotation.

### CODE REVIEW

File: src/esp/nav/GreedyFollower.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include "esp/nav/PathFinder.h"
+#include "esp/scene/SceneGraph.h"
+#include "esp/scene/SceneNode.h"
+
+namespace esp {
+namespace nav {
+
+class GreedyGeodesicFollowerImpl {
+ public:
+  enum class CODES { ERROR = -2, STOP = -1, FORWARD = 0, LEFT = 1, RIGHT = 2 };
```
Comment (msavva) [inline src/esp/nav/GreedyFollower.h]: short docstring for readability

File: src/esp/nav/GreedyFollower.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  typedef std::function<void(scene::SceneNode*)> MoveFn;
```
Comment (msavva) [inline src/esp/nav/GreedyFollower.h]: short docstring for readability

File: src/esp/nav/GreedyFollower.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  typedef std::tuple<vec3f, quatf> State;
```
Comment (msavva) [inline src/esp/nav/GreedyFollower.h]: short docstring for readability

File: src/esp/nav/GreedyFollower.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                             double turnAmount)
+      : pathfinder_{pathfinder},
+        moveForward_{moveForward},
+        turnLeft_{turnLeft},
+        turnRight_{turnRight},
+        goalDist_{goalDist},
+        forwardAmount_{forwardAmount},
+        turnAmount_{turnAmount} {};
+
+  int nextActionAlong(const State& start, const Eigen::Ref<const vec3f> end);
```
Comment (msavva) [inline src/esp/nav/GreedyFollower.h]: docstring with parameters

File: src/esp/nav/GreedyFollower.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+  inline int nextActionAlong(const Eigen::Ref<const vec3f> startPos,
```
Comment (msavva) [inline src/esp/nav/GreedyFollower.h]: docstring with parameters

File: src/esp/nav/GreedyFollower.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                             const Eigen::Ref<const vec4f> startRot,
+                             const Eigen::Ref<const vec3f> end) {
+    quatf rot = Eigen::Map<const quatf>(startRot.data());
+    return nextActionAlong(std::make_tuple(startPos, rot), end);
+  }
+
+  std::vector<int> findPath(const State& start,
```
Comment (msavva) [inline src/esp/nav/GreedyFollower.h]: docstring with parameters

File: src/esp/nav/GreedyFollower.h [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                            const Eigen::Ref<const vec3f> end);
+
+  inline std::vector<int> findPath(const Eigen::Ref<const vec3f> startPos,
```
Comment (msavva) [inline src/esp/nav/GreedyFollower.h]: docstring with parameters

File: src/esp/nav/GreedyFollower.h [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+namespace nav {
+
+class GreedyGeodesicFollowerImpl {
+ public:
+  enum class CODES { ERROR = -2, STOP = -1, FORWARD = 0, LEFT = 1, RIGHT = 2 };
+
+  typedef std::function<void(scene::SceneNode*)> MoveFn;
+  typedef std::tuple<vec3f, quatf> State;
+
+  GreedyGeodesicFollowerImpl(PathFinder::ptr& pathfinder,
```
Comment (msavva) [inline src/esp/nav/GreedyFollower.h]: docstring with parameters (even though this is Impl, it is in the header so good to have some docs for readability and extensibility)

File: tests/test_greedy_follower.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+import habitat_sim
+import habitat_sim.bindings as hsim
+import pytest
+import tqdm
+
+base_dir = osp.abspath(osp.join(osp.dirname(__file__), ".."))
+
+test_navmeshes = [
+    osp.join(base_dir, "data/scene_datasets/mp3d/17DRP5sb8fy/17DRP5sb8fy.navmesh"),
```
Comment (msavva) [inline tests/test_greedy_follower.py]: This mp3d scene isn't in the test scenes archive, should we keep it, or just use the other ones for this test?

Comment (erikwijmans) [reply tests/test_greedy_follower.py]: The test will skip if it doesn't find the navmesh :)

File: habitat_sim/simulator.py [hunk 10]
```diff
@@ -24,9 +26,9 @@ class Configuration(object):
 @attr.s
 class Simulator:
     config: Configuration = attr.ib()
-    agents: List = attr.ib(factory=list, init=False)
     _sim: hsim.SimulatorBackend = attr.ib(default=None, init=False)
     _num_total_frames: int = attr.ib(default=0, init=False)
+    agents: List[Agent] = attr.ib(factory=list, init=False)
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_sim/simulator.py]: Who knows!  Moved it back :)

**Comment by erikwijmans** [comment]
This PR provides a much faster and reliable greedy follower. While it doesn't generate the shortest path in action space, the paths still tend to be very good.

**Pull request merged**